app/components/combined_system_message/combined_system_message.js

```diff
@@ -0,0 +1,350 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import PropTypes from 'prop-types';
+import React from 'react';
+import {Text} from 'react-native';
+import {intlShape} from 'react-intl';
+
+import {Posts} from 'mattermost-redux/constants';
+import {displayUsername} from 'mattermost-redux/utils/user_utils';
+
+import CustomPropTypes from 'app/constants/custom_prop_types';
+
+import {makeStyleSheetFromTheme} from 'app/utils/theme';
+
+import FormattedText from 'app/components/formatted_text';
+
+import LastUsers from './last_users';
+
+const {
+    JOIN_CHANNEL, ADD_TO_CHANNEL, REMOVE_FROM_CHANNEL, LEAVE_CHANNEL,
+    JOIN_TEAM, ADD_TO_TEAM, REMOVE_FROM_TEAM, LEAVE_TEAM,
+} = Posts.POST_TYPES;
+
+const postTypeMessage = {
+    [JOIN_CHANNEL]: {
+        one: {
+            id: ['combined_system_message.first_user', 'combined_system_message.joined_channel'],
+            defaultMessage: ['{firstUser} ', 'joined the channel'],
```

Comment:
I'm a bit worried about constructing the message like this since it may cause problems in other languages, but I can't come up with a better way to do it with how they're formatted.

Reply:
I don't think there is a good way to do that in RN. The only thing I could think would be to maybe go in the direction of nesting the strings like
```
message = (
    <FormattedText
        defaultMessage='{firstUser} {action}'
        values={{
            action: (
                <FormattedText
                    style={{fontWeight: 'bold'}}
                    defaultMessage='joined the channel'
                />
            )
        }}
    />
);
```

That *may* be better as long as translators understand how to do it that way. Plus, we could reuse those strings between both apps

Reply:
That's neat and I like how it's nested. However, I'm 1/5 of not changing current implementation since it's still breaking the locale strings. I'm hoping that the alternative would have a one straight locale string per case. 
But I would definitely change the implementation just in case you have strong opinion for that.